### PR TITLE
HDDS-9019. ECUnderReplicationHandler rejects a potential target if it doesn't satisfy placement policy

### PR DESCRIPTION
**Pull request by siddhantsangwan**
## What changes were proposed in this pull request?

`ECUnderReplicationHandler` checks if current available source datanodes + potential replica datanodes satisfy placement policy. If they don't, the targets are rejected and recovery fails. While this check makes some sense, it can block container recovery. Since under replication is more serious than mis replication, we should let a new replica be reconstructed even if it doesn't satisfy placement policy. 

Also, we should rely on the policy to suggest a node that satisfies the policy - if there is none, the policy will fallback and try to suggest a node on a rack where the container is already present. 

Changes:
Allow recovery even if placement policy is not satisfied. Also added some logs.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9019

## How was this patch tested?

Modified existing UT according to the new expectation.

### CODE REVIEW

**Review by adoroszlai — APPROVED** [review]
Thanks @siddhantsangwan for the patch.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java [hunk 1]
```diff
@@ -334,8 +335,19 @@ private int processMissingIndexes(
...
+          validatePlacement(container, availableSourceNodes, selectedDatanodes);
+      if (!placementStatusWithSelectedTargets.isPolicySatisfied()) {
+        LOG.debug("Target nodes + existing nodes for EC container {}" +
+                " will not satisfy placement policy {}. Reason: {}. Selected" +
+                " nodes: {}. Available source nodes: {}. Resuming " +
+                "reconstruction regardless.",
+            container.containerID(), containerPlacement.getClass().getName(),
+            placementStatusWithSelectedTargets.misReplicatedReason(),
+            selectedDatanodes, availableSourceNodes);
+      }
```
[the remark this answers is not in the document]
Comment (siddhantsangwan) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java:349]: Good point... actually I had thought about this. I decided to let the caller of `validatePlacement()` control what it wants to do with the returned `ContainerPlacementStatus`. In this case that's logging "... Resuming reconstruction regardless."

[the remark this answers is not in the document]
Comment (adoroszlai) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java:349]: We may extract it into a separate method that just logs, but it's probably not worth the effort, given all the parameters that need to be passed.

**Comment by siddhantsangwan** [comment]
@adoroszlai thanks for the quick review.

**Pull request merged**